File: docs/scripts/verify_documentation_browser.py
"""Exercise a built documentation site with Playwright and a local Chrome binary.

Supply DOCUMENTATION_URL, CHROME_EXECUTABLE, and DOCUMENTATION_SCREENSHOTS.
The browser uses a fresh temporary profile; no personal browser data is read.
"""

import os
import json
import re
from pathlib import Path
from playwright.sync_api import sync_playwright, expect


address = os.environ.get("DOCUMENTATION_URL", "http://127.0.0.1:4321")
screenshots = Path(os.environ["DOCUMENTATION_SCREENSHOTS"])
screenshots.mkdir(parents=True, exist_ok=True)

with sync_playwright() as playwright:
    browser = playwright.chromium.launch(
        executable_path=os.environ["CHROME_EXECUTABLE"],
        chromium_sandbox=True,
    )
    try:
        context = browser.new_context(viewport={"width": 1440, "height": 1000})
        page = context.new_page()
        errors = []
        page.on("pageerror", lambda error: errors.append(str(error)))
        response = page.goto(address + "/")
        assert response.status == 200
        expect(page.locator("h1")).to_be_visible()
        documentation_paths = page.locator("#sidebar .navigation-group a").evaluate_all(
            "links => links.map(link => link.getAttribute('href'))")
        page.screenshot(path=str(screenshots / "documentation-desktop.png"))
        page.goto(address + "/getting-started/installation/")
        expect(page.locator("h1")).to_be_visible()
        expect(page.locator(".text-export")).to_have_attribute(
            "href", "/text/getting-started/installation/")
        expect(page.locator("#table-of-contents a").first).to_be_visible()
        expect(page.locator("code.language-sh .syntax-command").first).to_be_visible()
        original_examples = page.evaluate("""async () => {
          const html = await (await fetch(location.href)).text();
          const document = new DOMParser().parseFromString(html, 'text/html');
          return Array.from(document.querySelectorAll('pre > code'), code => code.textContent);
        }""")
        assert page.locator("pre > code").all_text_contents() == original_examples
        # Capture writes from the actual copy button without reading the host clipboard.
        page.evaluate("""() => {
          window.copiedDocumentation = null;
          Object.defineProperty(navigator.clipboard, 'writeText', {
            value: async (text) => { window.copiedDocumentation = text; }
          });
        }""")
        page.locator(".copy-button").first.click()
        page.wait_for_function("window.copiedDocumentation?.includes('nix')")
        assert page.evaluate("window.copiedDocumentation") == original_examples[0]
        theme = page.locator("#theme-select")
        theme.select_option("dark")
        expect(page.locator("html")).to_have_attribute("data-theme", "dark")
        page.reload()
        expect(page.locator("html")).to_have_attribute("data-theme", "dark")
        theme.select_option("light")
        expect(page.locator("html")).to_have_attribute("data-theme", "light")
        page.screenshot(path=str(screenshots / "documentation-installation-light.png"))
        page.locator("h1").click()
        page.keyboard.press("Control+k")
        expect(page.locator("#search-input")).to_be_focused()
        page.locator("h1").click()
        page.keyboard.press("Meta+k")
        expect(page.locator("#search-input")).to_be_focused()
        expect(page.get_by_role("navigation", name="Adjacent pages")).to_be_visible()
        expect(page.get_by_role("link", name="Report a documentation issue")).to_have_attribute(
            "href", "https://github.com/digitallyinduced/haskell-agent/issues/new")
        page.locator("#search-input").fill("sessions")
        page.locator("#search-input").press("Enter")
        results = page.locator(".search-results a")
        expect(results.first).to_be_visible()
        page.screenshot(path=str(screenshots / "documentation-search.png"))
        results.first.click()
        expect(page).to_have_url(address + "/guides/sessions/")
        expect(page.locator("h1")).to_be_visible()
        page.set_viewport_size({"width": 390, "height": 844})
        json_examples = 0
        for route in documentation_paths:
            response = page.goto(address + route)
            assert response.status == 200
            expect(page.locator("h1")).to_be_visible()
            assert page.evaluate("document.documentElement.scrollWidth <= window.innerWidth"), route
            for illustration in page.locator("article img").all():
                expect(illustration).to_have_attribute("alt", re.compile(r".+"))
                illustration.scroll_into_view_if_needed()
                page.wait_for_function("Array.from(document.querySelectorAll('article img')).every(image => image.complete && image.naturalWidth > 0)")
            for example in page.locator("pre > code.language-json").all_text_contents():
                json.loads(example)
                json_examples += 1
            headings = page.locator("article h2[id], article h3[id]").evaluate_all(
                "headings => headings.map(heading => '#' + heading.id)")
            contents = page.locator("#table-of-contents a").evaluate_all(
                "links => links.map(link => link.getAttribute('href'))")
            assert contents == headings, route
        assert json_examples > 0
        page.goto(address + "/reference/configuration/")
        page.locator(".mobile-contents summary").click()
        expect(page.locator(".mobile-contents a").first).to_be_visible()
        page.screenshot(path=str(screenshots / "documentation-configuration-mobile.png"))
        page.goto(address + "/")
        page.screenshot(path=str(screenshots / "documentation-mobile.png"))
        assert page.evaluate("document.documentElement.scrollWidth <= window.innerWidth")
        menu = page.locator("#menu-toggle")
        menu.click()
        expect(menu).to_have_attribute("aria-expanded", "true")
        page.keyboard.press("Escape")
        expect(menu).to_have_attribute("aria-expanded", "false")
        menu.click()
        page.locator("#sidebar a[href='/guides/sessions/']").click()
        expect(page).to_have_url(address + "/guides/sessions/")
        expect(menu).to_have_attribute("aria-expanded", "false")
        assert page.evaluate("document.documentElement.scrollWidth <= window.innerWidth")
        # The application must remain navigable and searchable without JavaScript.
        unenhanced = browser.new_context(
            java_script_enabled=False, viewport={"width": 390, "height": 844})
        unenhanced_page = unenhanced.new_page()
        unenhanced_page.goto(address + "/")
        unenhanced_page.locator(".mobile-contents summary").click()
        expect(unenhanced_page.locator(".mobile-contents a").first).to_be_visible()
        unenhanced_page.locator("#sidebar a[href='/guides/sessions/']").click()
        expect(unenhanced_page).to_have_url(address + "/guides/sessions/")
        unenhanced_page.locator("#search-input").fill("skills")
        unenhanced_page.locator("#search-input").press("Enter")
        expect(unenhanced_page.locator(".search-results a").first).to_be_visible()
        unenhanced.close()
        assert not errors, errors
        print(f"Browser checks passed: {len(documentation_paths)} pages, {json_examples} JSON examples, mobile overflow, section navigation, server search, themes, copy button, text export, and JavaScript-disabled use.")
    finally:
        browser.close()

File: docs/scripts/VerifyDocumentationBrowser.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ScopedTypeVariables #-}

-- Run with: nix develop .#docs -c runghc docs/scripts/VerifyDocumentationBrowser.hs
-- Uses Chrome's DevTools protocol directly, with a fresh temporary profile and
-- the normal browser sandbox. It never opens a personal browser profile.
import Control.Concurrent (threadDelay)
import Control.Exception.Safe (finally)
import Control.Monad (forM_, unless, void, when)
import Data.Aeson
import Data.Aeson.Key (fromText)
import Data.Aeson.Types (Pair)
import qualified Data.Aeson.KeyMap as KeyMap
import qualified Data.ByteString as ByteString
import qualified Data.ByteString.Base64 as Base64
import qualified Data.ByteString.Lazy as LazyByteString
import Data.IORef
import Data.Maybe (fromMaybe)
import Data.Text (Text)
import qualified Data.Text as Text
import qualified Data.Text.Encoding as Text
import qualified Network.HTTP.Client as HTTP
import qualified Network.WebSockets as WebSocket
import System.Directory
import System.Environment
import System.FilePath
import System.IO.Temp (withTempDirectory)
import System.Process
import System.Timeout (timeout)

data Browser = Browser WebSocket.Connection (IORef Int) (IORef [Value])

main :: IO ()
main = do
    address <- Text.pack . fromMaybe "http://127.0.0.1:4321" <$> lookupEnv "DOCUMENTATION_URL"
    executable <- getEnv "CHROME_EXECUTABLE"
    screenshots <- getEnv "DOCUMENTATION_SCREENSHOTS"
    temporaryRoot <- getEnv "TMPDIR"
    createDirectoryIfMissing True screenshots
    withTempDirectory temporaryRoot "documentation-browser-" $ \profile ->
        withCreateProcess (proc executable
            ["--headless=new", "--remote-debugging-port=0", "--remote-debugging-address=127.0.0.1",
             "--user-data-dir=" <> profile, "--no-first-run", "--no-default-browser-check",
             "--disable-background-networking", "--disable-component-update", "about:blank"])
            { std_out = NoStream } $ \_ _ _ process ->
            flip finally (terminateProcess process >> void (waitForProcess process)) $ do
                waitFor "Chrome debugging endpoint" (doesFileExist (profile </> "DevToolsActivePort"))
                endpoint <- lines <$> readFile (profile </> "DevToolsActivePort")
                port <- case endpoint of
                    value : _ -> pure (read value)
                    _ -> fail "Chrome did not publish a debugging port"
                manager <- HTTP.newManager HTTP.defaultManagerSettings
                request <- HTTP.parseRequest ("http://127.0.0.1:" <> show port <> "/json/new?about:blank")
                response <- HTTP.httpLbs request { HTTP.method = "PUT" } manager
                target <- either fail pure (eitherDecode (HTTP.responseBody response))
                identifier <- stringAt "id" target
                WebSocket.runClient "127.0.0.1" port ("/devtools/page/" <> Text.unpack identifier) $ \connection -> do
                    browser <- Browser connection <$> newIORef 0 <*> newIORef []
                    void $ command browser "Page.enable" []
                    void $ command browser "Runtime.enable" []
                    void $ command browser "Network.enable" []
                    verify browser address screenshots

command :: Browser -> Text -> [Pair] -> IO Value
command (Browser connection nextIdentifier errors) method parameters = do
    identifier <- atomicModifyIORef' nextIdentifier (\value -> (value + 1, value + 1))
    WebSocket.sendTextData connection (encode (object ["id" .= identifier, "method" .= method, "params" .= object parameters]))
    result <- timeout 15000000 (receive identifier)
    maybe (fail ("Timed out: " <> Text.unpack method)) pure result
  where
    receive identifier = do
        message <- WebSocket.receiveData connection :: IO LazyByteString.ByteString
        value <- either fail pure (eitherDecode message)
        when (field "method" value == String "Runtime.exceptionThrown") $
            modifyIORef' errors (value :)
        if field "id" value == toJSON identifier
            then if field "error" value /= Null
                then fail ("DevTools error: " <> show value)
                else pure (field "result" value)
            else receive identifier

field :: Text -> Value -> Value
field name (Object values) = fromMaybe Null (KeyMap.lookup (fromText name) values)
field _ _ = Null

stringAt :: Text -> Value -> IO Text
stringAt name value = case field name value of
    String result -> pure result
    other -> fail ("Expected text field " <> Text.unpack name <> ": " <> show other)

evaluate :: Browser -> Text -> IO Value
evaluate browser expression = do
    result <- command browser "Runtime.evaluate"
        ["expression" .= expression, "returnByValue" .= True, "awaitPromise" .= True]
    unless (field "exceptionDetails" result == Null) $
        fail ("Browser evaluation failed: " <> show result)
    pure (field "value" (field "result" result))

assertBrowser :: Browser -> Text -> IO ()
assertBrowser browser expression = do
    value <- evaluate browser expression
    unless (value == Bool True) $ fail ("Browser assertion failed: " <> Text.unpack expression <> ": " <> show value)

waitFor :: String -> IO Bool -> IO ()
waitFor description condition = do
    result <- timeout 15000000 loop
    unless (result == Just ()) $ fail ("Timed out waiting for " <> description)
  where
    loop = do
        ready <- condition
        unless ready (threadDelay 50000 >> loop)

waitBrowser :: Browser -> Text -> IO ()
waitBrowser browser expression = waitFor (Text.unpack expression) ((== Bool True) <$> evaluate browser expression)

literal :: Text -> Text
literal = Text.decodeUtf8 . LazyByteString.toStrict . encode

selector :: Text -> Text
selector value = "document.querySelector(" <> literal value <> ")"

visible :: Browser -> Text -> IO ()
visible browser value = waitBrowser browser $
    "(()=>{const e=" <> selector value <> ";if(!e)return false;const r=e.getBoundingClientRect();return r.width>0&&r.height>0&&getComputedStyle(e).visibility!=='hidden'})()"

click :: Browser -> Text -> IO ()
click browser value = do
    visible browser value
    position <- evaluate browser $ "(()=>{const e=" <> selector value <> ";e.scrollIntoView({block:'center'});const r=e.getBoundingClientRect();return {x:r.x+r.width/2,y:r.y+r.height/2}})()"
    forM_ ["mousePressed", "mouseReleased" :: Text] $ \event ->
        void $ command browser "Input.dispatchMouseEvent"
            ["type" .= event, "x" .= field "x" position, "y" .= field "y" position,
             "button" .= ("left" :: Text), "clickCount" .= (1 :: Int)]

key :: Browser -> Text -> Int -> IO ()
key browser name modifiers = forM_ ["keyDown", "keyUp" :: Text] $ \event ->
    void $ command browser "Input.dispatchKeyEvent"
        ["type" .= event, "key" .= name, "modifiers" .= modifiers,
         "text" .= (if event == "keyDown" && name == "Enter" then "\r" else "" :: Text),
         "windowsVirtualKeyCode" .= (if name == "Enter" then 13 else if name == "Escape" then 27 else 75 :: Int)]

navigate :: Browser -> Text -> IO ()
navigate browser url = do
    result <- command browser "Page.navigate" ["url" .= url]
    unless (field "errorText" result == Null) $ fail ("Navigation failed: " <> show result)
    waitBrowser browser ("location.href===" <> literal url <> "&&document.readyState==='complete'")
    -- Resource Timing records the actual navigation response, not a second fetch.
    assertBrowser browser "performance.getEntriesByType('navigation')[0].responseStatus===200"

viewport :: Browser -> Int -> Int -> IO ()
viewport browser width height = void $ command browser "Emulation.setDeviceMetricsOverride"
    ["width" .= width, "height" .= height, "deviceScaleFactor" .= (1 :: Int), "mobile" .= False]

screenshot :: Browser -> FilePath -> IO ()
screenshot browser path = do
    result <- command browser "Page.captureScreenshot" ["format" .= ("png" :: Text)]
    encoded <- stringAt "data" result
    content <- either fail pure (Base64.decode (Text.encodeUtf8 encoded))
    ByteString.writeFile path content

search :: Browser -> Text -> IO ()
search browser query = do
    click browser "#search-input"
    void $ command browser "Input.insertText" ["text" .= query]
    key browser "Enter" 0
    visible browser ".search-results a"

verify :: Browser -> Text -> FilePath -> IO ()
verify browser@(Browser _ _ errors) address screenshots = do
    viewport browser 1440 1000
    navigate browser (address <> "/")
    visible browser "h1"
    pathsValue <- evaluate browser "Array.from(document.querySelectorAll('#sidebar .navigation-group a'),a=>a.getAttribute('href'))"
    paths <- case fromJSON pathsValue of
        Success values -> pure (values :: [Text])
        Error message -> fail message
    unless (not (null paths)) $ fail "No documentation routes found"
    screenshot browser (screenshots </> "documentation-desktop.png")
    navigate browser (address <> "/getting-started/installation/")
    visible browser "h1"
    assertBrowser browser "document.querySelector('.text-export').getAttribute('href')==='/text/getting-started/installation/'"
    visible browser "#table-of-contents a"
    visible browser "code.language-sh .syntax-command"
    assertBrowser browser "(async()=>{const html=await(await fetch(location.href)).text();const original=new DOMParser().parseFromString(html,'text/html');window.originalDocumentationExamples=Array.from(original.querySelectorAll('pre > code'),e=>e.textContent);return JSON.stringify(Array.from(document.querySelectorAll('pre > code'),e=>e.textContent))===JSON.stringify(window.originalDocumentationExamples)})()"
    void $ evaluate browser "window.copiedDocumentation=null;Object.defineProperty(navigator.clipboard,'writeText',{value:async text=>{window.copiedDocumentation=text}})"
    click browser ".copy-button"
    waitBrowser browser "window.copiedDocumentation===window.originalDocumentationExamples[0]&&window.copiedDocumentation.includes('nix')"
    setTheme "dark"
    navigate browser (address <> "/getting-started/installation/")
    assertBrowser browser "document.documentElement.dataset.theme==='dark'"
    setTheme "light"
    screenshot browser (screenshots </> "documentation-installation-light.png")
    forM_ [2, 4] $ \modifier -> do
        click browser "h1"
        key browser "k" modifier
        assertBrowser browser "document.activeElement===document.querySelector('#search-input')"
    visible browser "nav[aria-label='Adjacent pages']"
    assertBrowser browser "Array.from(document.querySelectorAll('a')).some(a=>a.textContent==='Report a documentation issue'&&a.href==='https://github.com/digitallyinduced/haskell-agent/issues/new')"
    search browser "sessions"
    screenshot browser (screenshots </> "documentation-search.png")
    click browser ".search-results a"
    waitBrowser browser ("location.href===" <> literal (address <> "/guides/sessions/") <> "&&document.readyState==='complete'")
    visible browser "h1"
    viewport browser 390 844
    counts <- mapM (\route -> do
        navigate browser (address <> route)
        visible browser "h1"
        assertBrowser browser "document.documentElement.scrollWidth<=innerWidth"
        assertBrowser browser "Array.from(document.querySelectorAll('article img')).every(i=>i.alt.length>0)"
        void $ evaluate browser "document.querySelectorAll('article img').forEach(i=>i.loading='eager')"
        waitBrowser browser "Array.from(document.querySelectorAll('article img')).every(i=>i.complete&&i.naturalWidth>0)"
        assertBrowser browser "JSON.stringify(Array.from(document.querySelectorAll('article h2[id],article h3[id]'),e=>'#'+e.id))===JSON.stringify(Array.from(document.querySelectorAll('#table-of-contents a'),e=>e.getAttribute('href')))"
        count <- evaluate browser "Array.from(document.querySelectorAll('pre > code.language-json'),e=>JSON.parse(e.textContent)).length"
        case fromJSON count of
            Success value -> pure (value :: Int)
            Error message -> fail message) paths
    unless (sum counts > 0) $ fail "No JSON examples found"
    navigate browser (address <> "/reference/configuration/")
    click browser ".mobile-contents summary"
    visible browser ".mobile-contents a"
    screenshot browser (screenshots </> "documentation-configuration-mobile.png")
    navigate browser (address <> "/")
    screenshot browser (screenshots </> "documentation-mobile.png")
    assertBrowser browser "document.documentElement.scrollWidth<=innerWidth"
    click browser "#menu-toggle"
    assertBrowser browser "document.querySelector('#menu-toggle').getAttribute('aria-expanded')==='true'"
    key browser "Escape" 0
    assertBrowser browser "document.querySelector('#menu-toggle').getAttribute('aria-expanded')==='false'"
    click browser "#menu-toggle"
    click browser "#sidebar a[href='/guides/sessions/']"
    waitBrowser browser ("location.href===" <> literal (address <> "/guides/sessions/") <> "&&document.readyState==='complete'")
    assertBrowser browser "document.querySelector('#menu-toggle').getAttribute('aria-expanded')==='false'&&document.documentElement.scrollWidth<=innerWidth"
    -- DevTools evaluation remains available while page script execution is disabled.
    -- Interactions below use browser input events, not JavaScript event handlers.
    void $ command browser "Emulation.setScriptExecutionDisabled" ["value" .= True]
    navigate browser (address <> "/")
    assertBrowser browser "document.querySelector('.copy-button')===null"
    click browser ".mobile-contents summary"
    visible browser ".mobile-contents a"
    click browser "#sidebar a[href='/guides/sessions/']"
    waitBrowser browser ("location.href===" <> literal (address <> "/guides/sessions/") <> "&&document.readyState==='complete'")
    search browser "skills"
    recorded <- readIORef errors
    unless (null recorded) $ fail ("Browser page errors: " <> show recorded)
    putStrLn ("Browser checks passed: " <> show (length paths) <> " pages, " <> show (sum counts) <>
        " JSON examples, mobile overflow, section navigation, server search, themes, copy button, text export, and JavaScript-disabled use.")
  where
    setTheme theme = do
        void $ evaluate browser ("document.querySelector('#theme-select').value=" <> literal theme <> ";document.querySelector('#theme-select').dispatchEvent(new Event('change',{bubbles:true}))")
        assertBrowser browser ("document.documentElement.dataset.theme===" <> literal theme)
